subsys/bluetooth/host/att.c

```diff
@@ -2291,33 +2690,40 @@ void bt_att_req_free(struct bt_att_req *req)
 int bt_att_send(struct bt_conn *conn, struct net_buf *buf, bt_conn_tx_cb_t cb,
 		void *user_data)
 {
+	struct bt_att_chan *chan, *tmp;
 	struct bt_att *att;
-	int err;
+	int ret;
 
 	__ASSERT_NO_MSG(conn);
 	__ASSERT_NO_MSG(buf);
 
-	att = att_chan_get(conn);
+	att = att_get(conn);
 	if (!att) {
 		net_buf_unref(buf);
 		return -ENOTCONN;
 	}
 
-	/* Don't use tx_sem if caller has set it own callback */
-	if (!cb) {
-		/* Queue buffer to be send later */
-		if (k_sem_take(&att->tx_sem, K_NO_WAIT) < 0) {
-			k_fifo_put(&att->tx_queue, buf);
-			return 0;
-		}
+	/* If callback is set use the fixed channel since bt_l2cap_chan_send
+	 * cannot be used with a custom user_data.
+	 */
+	if (cb) {
+		bt_l2cap_send_cb(conn, BT_L2CAP_CID_ATT, buf, cb, user_data);
+		return 0;
 	}
 
-	err = att_send(conn, buf, cb, user_data);
-	if (err) {
-		if (!cb) {
-			k_sem_give(&att->tx_sem);
+	ret = 0;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&att->chans, chan, tmp, node) {
+		ret = bt_att_chan_send(chan, buf, NULL);
+		if (ret >= 0) {
+			break;
 		}
-		return err;
+	}
+
+	if (ret < 0) {
+		/* Queue buffer to be send later */
+		BT_DBG("Queueing buffer %p", buf);
+		k_fifo_put(&att->tx_queue, buf);
 	}
 
 	return 0;
```

Comment:
Should this now return `ret`?

Reply:
Nope, it is queueing if there is an error.